daemon/service-db.cc

```diff
@@ -298,15 +319,105 @@ MLServiceDB::set_model (const std::string name, const std::string model)
-      key_with_prefix.c_str (), key_with_prefix.c_str (), model.c_str ());
+  /* (key, version, active, path, description) */
+  sql = g_strdup_printf ("INSERT OR REPLACE INTO tblModel VALUES ('%s', IFNULL ((SELECT version from tblModel WHERE key = '%s' ORDER BY version DESC LIMIT 1) + 1, 1), '%s', '%s', '%s');",
+      key_with_prefix.c_str (), key_with_prefix.c_str (), is_active ? "T" : "F", model.c_str (), description.c_str ());
+  // rc = sqlite3_exec (_db, sql, _set_model_cb, nullptr, &errmsg);
   rc = sqlite3_exec (_db, sql, nullptr, nullptr, &errmsg);
   g_free (sql);
   if (rc != SQLITE_OK) {
-    g_warning ("Failed to insert pipeline description with name %s: %s (%d)", name.c_str (), errmsg, rc);
+    g_critical ("Failed to register model with name %s: %s (%d)", name.c_str (), errmsg, rc);
     sqlite3_clear_errmsg (errmsg);
-    throw std::runtime_error ("Failed to insert model.");
+    throw std::runtime_error ("Failed to register model.");
+  }
+
+  long long int last_id = sqlite3_last_insert_rowid (_db);
+  if (last_id == 0) {
+    g_critical ("Failed to get last inserted row id: %s", errmsg);
+    sqlite3_clear_errmsg (errmsg);
+    throw std::runtime_error ("Failed to get last inserted row id.");
+  }
+
+  /* get model's version */
+  sql = g_strdup_printf ("SELECT version FROM tblModel WHERE rowid = %lld ORDER BY version DESC LIMIT 1;", last_id);
+
+  /* callback method */
+  // rc = sqlite3_exec (_db, sql, _set_model_cb, version, &errmsg);
+
+  sqlite3_stmt *res;
+  rc = sqlite3_prepare_v2 (_db, sql, -1, &res, nullptr);
+
+  g_free (sql);
+  if (rc != SQLITE_OK) {
+    g_critical ("Failed to get model version with name %s: %s (%d)", name.c_str (), errmsg, rc);
+    sqlite3_clear_errmsg (errmsg);
+    throw std::runtime_error ("Failed to get model version.");
+  }
+
+  /* set rowid as the last_id int the sql */
+  sqlite3_bind_int (res, 1, last_id);
+
+  int step = sqlite3_step (res);
+  if (step == SQLITE_ROW) {
+    *version = sqlite3_column_int (res, 0);
+  }
+
+  sqlite3_finalize (res);
+
+  if (version == 0) {
+    g_critical ("Failed to get model version with name %s: %s (%d)", name.c_str (), errmsg, rc);
+    sqlite3_clear_errmsg (errmsg);
+    throw std::runtime_error ("Failed to get model version.");
+  }
+}
+
+void
+MLServiceDB::update_model_description (const std::string name, const guint version, const std::string description)
+{
+  int rc;
+  char *sql;
+  char *errmsg = nullptr;
+
+  if (name.empty ())
+    throw std::invalid_argument ("Invalid name parameter!");
+
+  std::string key_with_prefix = DB_KEY_PREFIX;
+  key_with_prefix += name;
+
+  /* (key, version, active, path, description) */
+  sql = g_strdup_printf ("UPDATE tblModel SET description = '%s' WHERE key = '%s' AND version = %u;",
+      description.c_str (), key_with_prefix.c_str (), version);
+  rc = sqlite3_exec (_db, sql, nullptr, nullptr, &errmsg);
```

Comment:
```SQL

-- If the result of the below query is '0', then return ML_ERROR_INVALID_PARAMETER Error 
SELECT count(*) FROM tblModel WHERE key = {key_with_prefix} AND version = {version};

UPDATE tblModel SET description = {description}
WHERE key = key_with_prefix AND version = {version};
```

Reply:
Thanks! I used the `SELECT EXISTS(SELECT 1 FROM tblModel WHERE key = '%s' AND version = %u);` to check the existence.